subsys/bluetooth/host/conn.c

```diff
@@ -768,6 +768,10 @@ void bt_conn_process_tx(struct bt_conn *conn)
 	buf = net_buf_get(&conn->tx_queue, K_NO_WAIT);
 	BT_ASSERT(buf);
 	if (!send_buf(conn, buf)) {
+		if (tx_data(buf)->tx) {
+			conn_tx_destroy(conn, tx_data(buf)->tx);
```

Comment:
After looking a bit more at the code I think you should also set tx to NULL after destroying it.
There are two more conditional calls to conn_tx_destroy() and those (conn_cleanup and send_frag) probably also should set tx to NULL...

Reply:
I looked at it too, but why should I do that ? is it so nothing uses the (destroyed) tx reference in the rest of the function (s) ? because the buffer will get destroyed and along with it its user data (where tx is stored).

Reply:
nvm, you're right: the failing test was because of this succession of events:
- `send_frag` failed, then called `conn_tx_destroy` but still kept the `tx` ref, because `always_consume` wasn't set.
- `bt_conn_process_tx` also failed, and attempted to destroy the (now invalid) `tx` ref -> segfault.

Reply:
The part about the callbacks in `conn_tx_destroy` still stands. It may be safer to detach `tx` first. We have adopted this pattern several places already.

```c
struct bt_conn_tx tx = tx_data(buf)->tx;
tx_data(buf)->tx = NULL;
conn_tx_destroy(conn, tx);
```

This also allows moving the `net_buf_unref` up. Freeing earlier is always better. (It's like releasing a lock.)
```c
struct bt_conn_tx detached_tx = NULL;
if (buf->ref == 1) {
	detached_tx = tx_data(buf)->tx;
	/* The following is not strictly needed because net_buf_unref invalidates `tx_data(buf)->tx`. */
	tx_data(buf)->tx = NULL;
}
net_buf_unref(buf);
if (detached_tx) {
	conn_tx_destroy(conn, tx);
}
```